### PR TITLE
Issue 5

### LINKED ISSUE #5: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [ ] When I add a new image, there is a form field to add tags.
- [ ] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [ ] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by sandipdev2020**
Closes #5, #6 , #7 
---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by sandipdev2020** [comment]
[Ropes Issue 4](https://trello.com/c/9ua0yYrx/1953-ropes-issue-4)

**Comment by sandipdev2020** [comment]
[Ropes Issue 6](https://trello.com/c/8R57JbUB/1955-ropes-issue-6)

**Comment by sandipdev2020** [comment]
[Ropes Issue 5](https://trello.com/c/OJRx7M5V/1954-ropes-issue-5)

File: Gemfile [hunk 1]
```diff
@@ -19,9 +19,12 @@ gem 'jquery-rails'
 
 gem 'webpacker', '~> 4'
 
-#simple form
+# Simple form
```
Comment (AgentCyrusJr) [inline Gemfile:22]: Very minor, it would be nice to commit this out of this commit, since this is not related to gem 'act-as-taggable-on'

File: test/models/image_test.rb [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 VALID_URL ||= 'https://www.gstatic.com/webp/gallery3/1.png'.freeze
 INVALID_URL ||= 'https://www.google.com'.freeze
+INVALID_URL_WITH_UNSUPPORTED_PROTOCOL ||= 'smtp://www.gstatic.com/webp/gallery3/1.png'.freeze
+INVALID_URL_WITH_UNSUPPORTED_FILE_EXTENSION ||= 'https://www.gstatic.com/webp/gallery3/1.dmg'.freeze
```
Comment (AgentCyrusJr) [inline test/models/image_test.rb:6]: I like this way of handling constant var

File: test/models/image_test.rb [hunk 3]
```diff
@@ -15,8 +17,24 @@ def test_image_with_empty_url
     assert_not_predicate @image, :valid?
   end
 
+  def test_image_with_invaid_url_protocol
```
Comment (AgentCyrusJr) [inline test/models/image_test.rb:20]: I am curious about why you would like to add these two tests `test_image_with_invaid_url_protocol` and `test_image_with_invalid_file_extension ` in this commit. Based on the change in `models/image.rb`, I would assume the validity of add this test `test_image_with_tag` since it tests on the presence of tags, however those two tests are for validation for url, which is not relevant to changes you made to the image model.

Also, I would suggest not commit the message as `modify image model`, since this message doesn't provide enough info to reviewers. How do you feel about something like `allow image to have valid tags` and `add tag association to image model`

File: app/views/images/show.html.erb [hunk 4]
```diff
@@ -1,8 +1,28 @@
...
-
-<%= link_to 'Back', images_path %>
+<div class="centered">
+  <h3> Imgage uploaded successfully</h3>
+  <table class="table">
+    <tr>
+      <td>
+        <strong>Image: <%= @image.id %></strong>
+        <br>
+        <%= image_tag(@image.url, :class => 'image-small') %>
```
Comment (AgentCyrusJr) [inline app/views/images/show.html.erb:8]: `class: 'image-small'` is the preferred syntax other than `:class => 'image-small'`

There is no functional difference between both, reference: https://stackoverflow.com/questions/8198811/whats-the-difference-between-colon-and-fat-arrow

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -21,10 +21,23 @@ def show
     @image = Image.find(params[:id])
   end
 
+  def tagged
```
Comment (AgentCyrusJr) [inline app/controllers/images_controller.rb:24]: Could you let me know the reason why we would like to introduce tag action in this controller?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -32,7 +34,7 @@ def test_new_image_path
   end
 
   def test_create_new_image_with_redirect
-    assert_difference('Image.count') do
+    assert_difference('Image.count', 1) do
```
Comment (AgentCyrusJr) [inline test/controllers/images_controller_test.rb:37]: `assert_difference` will make by default `1` assertion if no params are defined. 

reference: https://apidock.com/rails/ActiveSupport/Testing/Assertions/assert_difference

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -54,4 +56,46 @@ def test_view_image
...
+    assert_response :redirect
+    assert_redirected_to image_path(Image.last)
+    assert_includes Image.last.url, VALID_URL
+    follow_redirect!
+    assert_select 'a[href=?]', VALID_TAG
+  end
+
+  def  test_create_new_image_with_tags_and_jump_to_tag_page
+    Image.create!(url: VALID_URL, tag_list: 'flower')
+    get VALID_TAG
```
Comment (AgentCyrusJr) [inline test/controllers/images_controller_test.rb:73]: We can discuss about this together tomorrow, because I believe this is against my understanding on how to implement filter by tags

File: config/routes.rb [hunk 8]
```diff
@@ -2,6 +2,7 @@
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
   resources :images
   root 'images#index'
+  get '/tagged', to: 'images#tagged', as: :tagged
```
Comment (AgentCyrusJr) [inline config/routes.rb:5]: Same here, I expected tag is parsed as a param in `show` action
